docs(website): add date-picker doc page (#1498)
twilio-labs/paste: packages/paste-website/src/pages/components/date-picker/index.mdx

import {DatePicker, formatReturnDate} from '@twilio-paste/date-picker';
import {UnorderedList, ListItem} from '@twilio-paste/list';
import {Stack} from '@twilio-paste/stack';
import {Paragraph} from '@twilio-paste/paragraph';
import {Table, THead, TBody, Td, Th, Tr} from '@twilio-paste/table';
import {useUID} from '@twilio-paste/uid-library';
import {Combobox} from '@twilio-paste/combobox';
import {SidebarCategoryRoutes} from '../../../constants';
import {Callout, CalloutTitle, CalloutText} from '../../../components/callout';
import {
  RequiredDatePicker,
  DateRangePicker,
  OnChangePicker,
  HasErrorPicker,
  DisabledPicker,
  ReadOnlyPicker,
  LabelOnlyPicker,
  StackOfPickers,
  MinAndMaxPicker,
} from '../../../component-examples/DatePickerExamples.ts';

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/date-picker"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(fields: {slug: {eq: "/components/date-picker/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }

>     }

Author: After the new component overview table PR is merged and you rebase with main, you'll need to add this to the graphql query after the mdx part:
```
allAirtable(filter: {data: {Feature: {eq: "Date Picker"}}}) {
      edges {
        node {
          data {
            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
          }
        }
      }
    }
```

  }
`;

<ComponentHeader
  name="Date Picker"
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/date-picker"
  storybookUrl="/?path=/story/components-date-picker--default-date-picker"

>   data={props.data.allPasteComponent.edges}

Author: Also add `packageStatus={props.data.allAirtable.edges}` after this line.

Reviewer: Thanks! I'll make these changes after rebasing

/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

<Callout>
  <CalloutTitle>Using alpha components in your product</CalloutTitle>
  <CalloutText>
    Alpha components are ready for use in production and components like Date Picker won't likely see any API changes.
    The term "Alpha" means there is only basic documentation at this time. Please share your experience using this Alpha
    component by{' '}
    <Anchor href="https://github.com/twilio-labs/paste/discussions/new" showExternal>
      starting a discussion on GitHub
    </Anchor>
    .
  </CalloutText>

> </Callout>

Author: Feel free to delete this part. We're removing it on other pages in the stage 1, 2, ... PR.

## Guidelines

### About Date Picker

> The Date Picker should be used when a date-only value is required from the user. It is an input field which accepts a value in <inlineCode>MM/DD/YYYY</inlineCode> format.

Reviewer 2: > The Date Picker should be used when a date-only value is required from the user.

As a user reading this, `date-only` reads restrictive but it isn't really. I kind of expect a date picker to let me pick dates. 
After removing the `-only` the rest of the sentence doesn't really say anything meaningful so we can probably get rid of it and lead into the second sentence like:
`The Data Picker is an input field which accepts a value in ....`

Reviewer: That's a good point

The Date Picker is an input field which accepts a value in <inlineCode>MM/DD/YYYY</inlineCode> format.

Currently, the Date Picker is built on top of <Anchor href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date" showExternal>the HTML date picker</Anchor>, using an <inlineCode>input</inlineCode> field with <inlineCode>type="date"</inlineCode>, which comes with some inherent limitations.
In order to release a functional and stable Date Picker for immediate use, we decided to work within these limitations with the intention of updating this package to a custom Date Picker in a future release. The current API was designed to avoid any breaking changes with future versions of the Date Picker.

> ### Broswer support

Author: ```suggestion
### Browser support
```

> In order to release a functional and stable Date Picker for immediate usage, we decided to work with these limitations, with the intentions of updating this package to a custom-built Date Picker in a future release. The current API was designed to avoid releasing any breaking changes with future versions of the Date Picker.

Reviewer 2: ```suggestion
In order to release a functional and stable Date Picker for immediate use, we decided to work within these limitations with the intention of updating this package to a custom Date Picker in a future release. The current API was designed to avoid any breaking changes with future versions of the Date Picker.
```

        <Td>
          <Box
            margin="space20"
            as="img"
            src="/images/date-picker/firefox.png"
            alt="Chrome/Edge implementation of date picker"
            display="block"
            marginLeft="auto"
            marginRight="auto"
            maxWidth="300px"
            height="auto"
          />
        </Td>
      </Tr>
    </TBody>
  </Table>
</Box>

While cross-browser functionality, style and accessibility differences are present, they will not affect the usability of the Date Picker. The API is stable and full usage of the component is encouraged. There are certain use cases not included in the <inlineCode>input type="date"</inlineCode> functionality, such as a date range picker. You can find examples of how to recreate those use cases below. If there's a Date Picker use case not covered by the current implementation, feel free to <Anchor href="https://github.com/twilio-labs/paste/discussions/new" showExternal>open a Discussion</Anchor> so that we can help you find a solution.

### Accessibility

<UnorderedList>
  <ListItem>
    Include a visible label on <strong>all</strong> date pickers.
  </ListItem>
  <ListItem>

>     Each label must use the <inlineCode>htmlFor</inlineCode> prop that equals the <inlineCode>id</inlineCode> of the

Reviewer 2: ```suggestion
    Each label must use the <inlineCode>htmlFor</inlineCode> prop that matches the <inlineCode>id</inlineCode> of the
```

<LivePreview scope={{DatePicker, Label, HelpText, useUID, Anchor, Box, Combobox, formatReturnDate}} noInline>
  {OnChangePicker}
</LivePreview>

## States

### DatePicker with inline error

Change a date picker to its error state and add an inline error if the value entered doesn't pass validation requirements.

An inline error is placed below the field to inform a user of any errors in their value. If help text exists, the error text should replace and repeat the help text.

<LivePreview scope={{DatePicker, Label, HelpText, useUID, Anchor}} noInline>
  {HasErrorPicker}
</LivePreview>

### Disabled input

Use a disabled form field to show users that they can't interact with the field.

If you want to show information that can't be edited, use a [read-only](#read-only-input) form field or consider another way of showing static information.

<LivePreview scope={{DatePicker, Label, HelpText, useUID, Anchor}} noInline>
  {DisabledPicker}
</LivePreview>

### Read-only input

Use a read-only form field when a field's value can't be changed, but users should be able to read or apply focus on the field. For example, use a read-only form field if a user needs to copy the value.

> <LivePreview scope={{DatePicker, Label, HelpText, useUID, Anchor}} noInline>

Reviewer: Were you able to copy the date on any of the other examples? As far as I've seen, only one section of the date value (either `mm` or `dd` or `yyyy`) can be highlighted at a time. That's a really good point though and something to keep in mind to consider for the custom build!

  {ReadOnlyPicker}
</LivePreview>

## Composition notes

An input field must have at least a label and an input.

<LivePreview scope={{DatePicker, Label, useUID}} noInline>
  {LabelOnlyPicker}
</LivePreview>

<Callout>

>   <CalloutTitle as="h4">Quick note about Date Picker DOM methods</CalloutTitle>

Reviewer 2: 🔥

  <CalloutText>
    There are three methods that come with the HTML picker: <inlineCode>stepUp()</inlineCode>,{' '}
    <inlineCode>stepDown()</inlineCode> and <inlineCode>select()</inlineCode>. Usage of these methods is possible, but
    we <strong>do not</strong> encourage it, as they may not be supported in future versions of the Date Picker. Use at
    your own risk!
  </CalloutText>
</Callout>

### Positioning form fields

> Stack form fields vertically with `$space-80` spacing between each field.

Reviewer: Good catch. The wording is a little confusing here (copied from Input). I added `To stack form fields, you can use either a [Box](url) or a [Stack](url) (as seen in the example below).`

Don't prevent form submission by disabling the submit button. An error message can give more information than a disabled button can to help users figure out which fields are invalid.

Usage of the <inlineCode>min</inlineCode> and <inlineCode>max</inlineCode> props on Date Picker is encouraged if certain dates are invalid.

<LivePreview scope={{DatePicker, Label, HelpText, useUID, Anchor}} noInline>
  {MinAndMaxPicker}
</LivePreview>

### Errors

Use inline error text to explain how to fix an error.

Ideally, help text should have enough information to help users prevent errors. If help text exists and you need to show an error, the error text should replace and repeat the help text until the error has been resolved.

Error text should:

- Be actionable. Explain how to fix an error and if reasonable, why it happened so that it might also be prevented in the future.
- Be concise and simple, maybe even more than normal. Avoid jargon. Try to keep error text to 2 sentences or fewer.
- Use the passive voice for input errors to avoid placing blame on the user. For example, "A friendly name is required."

> - Use the active voice for system errors. For example, "Our systems are currently down. Please contact our support team."

Reviewer 3: Do you think including a code example like for the previous sections would be helpful here, or just overkill?

Reviewer: Hmm I'm leaning against the code example for this section because we already have the inline error example but could be convinced otherwise

### Optional compositional elements

- **Icon** — Use an icon if you need to give the user additional controls for the field. For example, use an icon to clear a date picker on click, removing the need for users to manually delete their entered value. Place icon buttons that trigger an action on the right side of the date picker. If you need 2 actions on a field (e.g., popover trigger and clear field), place the icon button that affects the date value on the right, and the other icon on the left.

> 

Reviewer 3: Do you think a code example here might help illustrate the guidance more clearly?

Reviewer: Yeah definitely, good catch

Reviewer: Ok Simon told me these are describing using [prefixes and suffixes](https://paste.twilio.design/components/input#input-with-add-ons-prefix-suffix-text-or-icons) (which we're not including for date picker) rather than a freestanding icon button. So scrapping this section anyways!

                <ListItem>Between label and required indicator: $space-20</ListItem>
              </UnorderedList>
            </ListItem>
            <ListItem>
              DatePicker:
              <UnorderedList marginBottom="space0">
                <ListItem>Left padding: $space-40</ListItem>
                <ListItem>Top, right, bottom padding: $space-30</ListItem>
              </UnorderedList>
            </ListItem>
            <ListItem>
              Help text:
              <UnorderedList marginBottom="space0">
                <ListItem>Top: $space-20</ListItem>
              </UnorderedList>
            </ListItem>
            <ListItem>
              Inline error:
              <UnorderedList marginBottom="space0">
                <ListItem>Spacing between icon and text: $space-20</ListItem>
                <ListItem>Top: $space-20</ListItem>
              </UnorderedList>
            </ListItem>
          </UnorderedList>
        </Td>
        <Td>No</Td>
      </Tr>
    </TBody>
  </Table>
</Box>

---

## Usage Guide

### API

#### Installation

```bash
yarn add @twilio-paste/date-picker - or - yarn add @twilio-paste/core
```

#### Usage

```jsx

> import {DatePicker, Label, HelpText, formatReturnDate} from '@twilio-paste/core';

Reviewer 2: Let's change the import line here to reference the unbarreled packages, like:
```
import {DatePIcker} from '@twilio-paste/core/date-picker';
etc
```
More deets: https://deploy-preview-1498--paste-docs.netlify.app/getting-started/engineering/#how-the-core-package-stays-lean

| onFocus?  | `(event: React.MouseEvent<HTMLElement>)` |                                                                      | null      |
| onBlur?   | `(event: React.MouseEvent<HTMLElement>)` |                                                                      | null      |
| variant?  | 'default', 'inverse'                     |                                                                      | 'default' |

#### formatReturnDate() Props

| Prop       | Type   | Description                                                                                                           | Default |
| ---------- | ------ | --------------------------------------------------------------------------------------------------------------------- | ------- |
| dateValue  | string | Date value to be formatted. Must be in <inlineCode>yyyy-mm-dd</inlineCode> format.                                    | null    |
| dateFormat | string | Speficy the format in which to return the dateValue. Formats come from the <inlineCode>date-fns</inlineCode> library. | null    |

> #### Label Props

Reviewer 3: Since Label and HelpText are separate components, would it make more sense to have these prop tables live on those doc pages? I can see that both those doc pages are empty, so this might be a topic for another day haha.

Reviewer: Yeah, I was talking to @SiTaggart about this recently. I think it comes from when `Form` was it's own package that exported Input, Label, HelpText, etc. Might make sense to leave it there for now (to be consistent with the Input page) until we rework the component pages. What do you think?
